### PR TITLE
Fix providing a `RoomStreamToken` instance to `_notify_app_services_ephemeral`

### PR DESCRIPTION
**Pull request by anoadragon453**
We typically figure out which EDUs to send to application services based the last stream token that appservice successfully received. This is true for read receipts and user presence, however for typing we always just send the last typing event. This process is started whenever a new read receipt, presence update or typing event is processed by the server. In that case, `notify_interested_services_ephemeral` will be called with the new event and its stream token (so we can check whether we need to send any other events to the appservice may have missed).

With all that in mind, I believe I've found a bug in the implementation. If a `RoomStreamToken` object was provided as the `new_token` argument to `_notify_app_services_ephemeral`, `new_token` would be set to `None`, before then being passed to `notify_interested_services_ephemeral` and then `_notify_interested_services_ephemeral`. The commit that introduced this change (559974fb4bc1872abced395de22ffbd2292b2f8b) was intended to prevent mypy from failing. `(_)notify_app_services_ephemeral` was also configured to allow `new_token` to be an `Optional[int]` as a result.

However, if `None` is provided, `_notify_app_services_ephemeral` will end up passing `None` to `set_type_stream_id_for_appservice`, which will clear each appservice's stream token for the given `stream_key`. ~~Doing so means that we send all presence or read receipt data from a None stream token up until the latest token(!). This seems like a bug to do every time we have a `RoomStreamToken` rather than an `int`.~~ Actually, all stream keys that are passed to `_notify_interested_services_ephemeral` (`typing_key`, `receipt_key`, `presence_key`) are passing `int` values for `new_token`, so the buggy code would have never fired anyways.

Regardless, I propose converting the `RoomStreamToken` to an `int` using `RoomStreamToken.stream` (~~I _think_ this is the right way to convert it. [`RoomStreamToken`](https://github.com/matrix-org/synapse/blob/aa2c027792d04c36b17866710e95a41d31f5d99c/synapse/types.py#L415-L464) can end up being a vector clock when you have multiple event persister workers; in this case `stream` will be the minimum stream token amongst all workers... which I think is fine... maybe?~~ This appears to be correct, we do it [elsewhere as well](https://github.com/matrix-org/synapse/blob/9abbd08e33fc7710d952e870e26368d17915bbba/synapse/handlers/appservice.py#L67-L70).)

~~This may relate to the large bursts of EDU traffic Synapse is sending to AS's ala https://github.com/matrix-org/synapse/issues/10836.~~ See above. The code should not have actually resulted in broken behaviour.

### CODE REVIEW

File: synapse/handlers/appservice.py [hunk 1]
```diff
@@ -212,6 +212,13 @@ def notify_interested_services_ephemeral(
         if stream_key not in ("typing_key", "receipt_key", "presence_key"):
             return
 
+        # Convert new_token from a RoomStreamToken to an int if necessary.
+        # We just use the minimum stream ordering and ignore the vector clock
+        # component. This is safe to do as long as we *always* ignore the vector
+        # clock components.
+        if isinstance(new_token, RoomStreamToken):
+            new_token = new_token.stream
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply synapse/handlers/appservice.py]: Good point -- I've changed this to an assert and left some context.

File: synapse/handlers/appservice.py [hunk 2]
```diff
@@ -212,12 +212,18 @@ def notify_interested_services_ephemeral(
...
             return
 
-        # Convert new_token from a RoomStreamToken to an int if necessary.
-        # We just use the minimum stream ordering and ignore the vector clock
-        # component. This is safe to do as long as we *always* ignore the vector
-        # clock components.
-        if isinstance(new_token, RoomStreamToken):
-            new_token = new_token.stream
+        # Assert that new_token is an integer (and not a RoomStreamToken).
+        # All of the supported streams that this function handles uses an
```
Comment (richvdh) [inline synapse/handlers/appservice.py:216]:
```suggestion
        # All of the supported streams that this function handles use an
```

File: synapse/handlers/appservice.py [hunk 3]
```diff
@@ -212,12 +212,18 @@ def notify_interested_services_ephemeral(
...
+        # integer to track progress (rather than a RoomStreamToken - a
+        # vector clock implementation - as they don't support multiple
+        # stream writers).
```
Comment (richvdh) [inline synapse/handlers/appservice.py:219]:
```suggestion
        # vector clock implementation) as they don't support multiple
        # stream writers.
```

**Review by richvdh — APPROVED** [review]
lgtm

**Pull request merged**